unmarshal.go

```diff
@@ -0,0 +1,336 @@
+package ipld
```

Comment:
Not this one.

- In `coding` you have the functions `Encode(...)` and `Decode(...)` that takes a `io.Reader` or `io.Writer`.
- In `stream` you have the `NodeReader` that can be marshalled and unmarshalled into arbitrary datatypes. The Go reflection is used here to perform the trick. It uses type annotation and is independant from the underlying coding (it is generic to the `NodeReader`)

Marshalling and unmarshalling do not transform a fixed binary representation to a structured type (that would be a codec). They take a generic node reader.